pkg/apis/traceflow/v1/types.go

```diff
@@ -0,0 +1,155 @@
+	INITIAL Phase = "INITIAL"
+	RUNNING Phase = "RUNNING"
+	SUCCESS Phase = "SUCCESS"
+	TIMEOUT Phase = "TIMEOUT"
+	ERROR   Phase = "ERROR"
+)
+
+type Component string
+
+const (
+	SPOOFGUARD    Component = "SPOOFGUARD"
+	LB            Component = "LB"
+	ROUTING       Component = "ROUTING"
+	NETWORKPOLICY Component = "NETWORKPOLICY"
+	FORWARDING    Component = "FORWARDING"
+)
+
+type Action string
+
+const (
+	DELIVERED Action = "DELIVERED"
+	RECEIVED  Action = "RECEIVED"
+	FORWARDED Action = "FORWARDED"
+	DROPPED   Action = "DROPPED"
+)
+
+type PacketTransportType string
+
+const (
+	UNICAST PacketTransportType = "UNICAST"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Source      `json:"source"`
+	Destination `json:"destination"`
+	RoundID     string `json:"roundID,omitempty"`
+
+	Packet `json:"packet"`
+	Status `json:"status"`
+}
+
+type Source struct {
+	Namespace string `json:"namespace,omitempty"`
+	Pod       string `json:"pod,omitempty"`
+}
+
+type Destination struct {
+	Namespace string `json:"namespace,omitempty"`
+	Pod       string `json:"pod,omitempty"`
+	Service   string `json:"service,omitempty"`
+	IP        string `json:"IP,omitempty"`
+}
+
+type IPHeader struct {
+	SrcIP    string `json:"srcIP,omitempty"`
+	Protocol int32  `json:"protocol,omitempty"`
+	TTL      int32  `json:"ttl,omitempty"`
+	Flags    int32  `json:"flags,omitempty"`
+}
+
+type TransportHeader struct {
+	ICMPEchoRequestHeader `json:"icmpEchoRequestHeader"`
+	UDPHeader             `json:"udpHeader"`
+	TCPHeader             `json:"tcpHeader"`
+}
+
+type ICMPEchoRequestHeader struct {
+	ID       int32 `json:"id,omitempty"`
+	Sequence int32 `json:"sequence,omitempty"`
+}
+
+type UDPHeader struct {
+	SrcPort int32 `json:"srcPort,omitempty"`
+	DstPort int32 `json:"dstPort,omitempty"`
+}
+
+type TCPHeader struct {
+	SrcPort int32 `json:"srcPort,omitempty"`
+	DstPort int32 `json:"dstPort,omitempty"`
+	Flags   int32 `json:"flags,omitempty"`
+}
+
+type Packet struct {
+	TransportType PacketTransportType `json:"transportType,omitempty"`
+	PayloadString string              `json:"payloadString,omitempty"`
+	PayloadSize   int32               `json:"payloadSize,omitempty"`
+
+	IPHeader        `json:"ipHeader"`
+	TransportHeader `json:"transportHeader"`
+}
+
+type Status struct {
+	Phase        Phase         `json:"phase,omitempty"`
+	CrossNodeTag uint8         `json:"crossNodeTag,omitempty"` // unique ID for traceflow
+	NodeSender   []Observation `json:"nodeSender,omitempty"`
+	NodeReceiver []Observation `json:"nodeReceiver,omitempty"`
+}
+
+type Observation struct {
+	Component       Component  `json:"component,omitempty"`
+	ComponentName   string     `json:"componentName,omitempty"`
+	Action          Action     `json:"action,omitempty"`
+	RoundID         string     `json:"roundID,omitempty"`
```

Comment:
Is this RoundID the same as the RoundID in Traceflow struct? If so why we need to duplicate it?

Reply:
Removed since we only have a sender and a receiver now.

Reply:
For Sender/ReceiverNode fields or Node slice, I will still wait for your answer whether feasible to use a slice. See my points above.
@gran-vmv

Reply:
I prefer to use Sender/ReceiverNode as different slice, because it is not safe to append to same slice from multiple clients.

Reply:
> I prefer to use Sender/ReceiverNode as different slice, because it is not safe to append to same slice from multiple clients.

This is not a problem, it's not allowed to update "SenderNode" and "ReceiverNode" concurrently either. Multiple clients that update the same CRD must fetch the latest version of the resource, apply their own change, and call the update API.
I think it makes difference between two ways.

Reply:
From the perspective of the extensibility, my original suggestion was having a slice of Observations.
@jianjuns is this also what you want? or you mean a slice of Nodes, each of which contains a slice of Observations?

Reply:
Discussed with team, let's change this to "observations"

Reply:
A slice of Observations is fine to me, but I thought according to @gran-vmv, a separate struct for each Node (it means a per Node struct like NodeObservations which includes Node info and a slice of observations) might help support updating from multiple Nodes?

Reply:
@jianjuns Like this? I feel OK for this change.
status:
  results:
  - node: node1
    timestamp: 123
    role: Sender
    observations:
    - component: xxx
    - component: xxx
  - node: node2
    timestamp: 124
    role: Receiver
    observations:
    - component: xxx
    - component: xxx

Reply:
I do not really have a preference over these two, but want to know which way is safe for parallel updates from multiple Nodes.

Reply:
All the two methods require RetryOnConflict when update.
The existing method will generate many duplicate properties such as node/timestamp in each observation, and the new method won't.

Reply:
Ok, I am fine with either way, as long as we have a way to solve the write conflicts.

Reply:
Updated with latest design.